pkg/typesystem/typesystem.go

```diff
@@ -667,163 +667,204 @@ func (t *TypeSystem) relationInvolvesExclusion(objectType, relation string, visi
-		v[typeName] = map[string]struct{}{
-			relationName: {},
-		}
+	if _, ok := v[typeName]; !ok {
+		// first time visiting this type
+		v[typeName] = map[string]*relationDetails{}
+	}
+	if _, ok := v[typeName][relationName]; !ok {
+		// first time visiting this relation
+		v[typeName][relationName] = &relationDetails{false, false}
 	}
-
 	relation, ok := typedefs[typeName][relationName]
 	if !ok {
-		return false, false, fmt.Errorf("undefined type definition for '%s#%s'", typeName, relationName)
+		return nil, fmt.Errorf("undefined type definition for '%s#%s'", typeName, relationName)
 	}
 
 	switch rw := rewrite.Userset.(type) {
 	case *openfgav1.Userset_This:
+		// At least one type must have an entrypoint.
 		for _, assignableType := range relation.GetTypeInfo().GetDirectlyRelatedUserTypes() {
+			assignableTypeName := assignableType.GetType()
+
+			if _, ok := typedefs[assignableTypeName]; !ok {
+				return nil, fmt.Errorf("undefined type '%s'", assignableTypeName)
+			}
 			if assignableType.GetRelationOrWildcard() == nil || assignableType.GetWildcard() != nil {
-				return true, false, nil
+				v[typeName][relationName].hasEntrypoints = true
+				break
 			}
 
-			assignableTypeName := assignableType.GetType()
 			assignableRelationName := assignableType.GetRelation()
 
 			assignableRelation, ok := typedefs[assignableTypeName][assignableRelationName]
 			if !ok {
-				return false, false, fmt.Errorf("undefined type definition for '%s#%s'", assignableTypeName, assignableRelationName)
+				return nil, fmt.Errorf("undefined type definition for '%s#%s'", assignableTypeName, assignableRelationName)
 			}
 
 			if _, ok := v[assignableTypeName][assignableRelationName]; ok {
+				// we entered a potential cycle, so keep on looking
 				continue
 			}
 
-			hasEntrypoint, _, err := hasEntrypoints(typedefs, assignableTypeName, assignableRelationName, assignableRelation.GetRewrite(), v)
+			relDetails, err := getRelationDetailsRecursive(typedefs, assignableTypeName, assignableRelationName, assignableRelation.GetRewrite(), v)
 			if err != nil {
-				return false, false, err
+				return nil, err
 			}
 
-			if hasEntrypoint {
-				return true, false, nil
+			if relDetails.hasEntrypoints {
+				v[typeName][relationName].hasEntrypoints = true
+				break
 			}
 		}
 
-		return false, false, nil
+		return v[typeName][relationName], nil
 	case *openfgav1.Userset_ComputedUserset:
 
 		computedRelationName := rw.ComputedUserset.GetRelation()
 		computedRelation, ok := typedefs[typeName][computedRelationName]
 		if !ok {
-			return false, false, fmt.Errorf("undefined type definition for '%s#%s'", typeName, computedRelationName)
+			return nil, fmt.Errorf("undefined type definition for '%s#%s'", typeName, computedRelationName)
 		}
 
-		if _, ok := v[typeName][computedRelationName]; ok {
-			return false, true, nil
+		if r, ok := v[typeName][computedRelationName]; ok {
+			// we entered a definitive cycle, so return
+			r.hasLoop = true
+			return r, nil
```

Comment:
Line 765 fixes #1385. If we found a loop, we return the correct value for `hasEntrypoints`, and not just `false`.

Reply:
Could you elaborate why this fixes #1385? I don't see a loop whatsoever in that example, so it's not clear to me why this logic herein would pertain to that specific issue.

At first glance it would seem that the visited map, specifically for the `but not` case would be more applicable.